🚧 [Hosts] Backup Settings

davidegiacometti opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #37666
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<img width="686" alt="image" src="https://github.com/user-attachments/assets/5e8ad75d-6286-4f0f-b210-21b346759499" />

<img width="686" alt="image" src="https://github.com/user-attachments/assets/ed952815-12e8-4efa-82fe-159c72c6da44" />

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

htcfreek commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -4895,6 +4895,41 @@ To record a specific window, enter the hotkey with the Alt key in the opposite m
+    <value>Specify the number of days to retain backups</value>
+  </data>
+  <data name="Hosts_DeleteBackupWarning.Title" xml:space="preserve">
+    <value>Backups will never will be kept indefinitely with the current configuration</value>
```
```suggestion
    <value>Backups will be kept indefinitely with the current configuration</value>
```

davidegiacometti replied:
Thanks for catching this (haven't reviewed the PR yet 😄).
What I wanted here is **Backups will be kept indefinitely with the current configuration**.

htcfreek replied:
The UI is confusing: If I keep my backups 0 days, then I have no backups. 🤔

htcfreek commented on src/settings-ui/Settings.UI/SettingsXAML/Views/HostsPage.xaml:
```diff
@@ -63,6 +63,83 @@
+                                    IsTabStop="{x:Bind ViewModel.ShowDeleteBackupWarning, Mode=OneWay}"
+                                    Severity="Warning" />
+                            </tkcontrols:SettingsCard>
+                            <tkcontrols:SettingsCard x:Uid="Hosts_Backup_DaysToKeep" IsEnabled="{x:Bind ViewModel.BackupHosts, Mode=OneWay}">
```
is enabled should be bind against both: backup hosts and automatic deletion

htcfreek commented on src/settings-ui/Settings.UI/SettingsXAML/Views/HostsPage.xaml:
```diff
@@ -63,6 +63,83 @@
+                                    SpinButtonPlacementMode="Compact"
+                                    Value="{x:Bind ViewModel.DaysToKeep, Mode=TwoWay}" />
+                            </tkcontrols:SettingsCard>
+                            <tkcontrols:SettingsCard x:Uid="Hosts_Backup_CopiesToKeep" IsEnabled="{x:Bind ViewModel.BackupHosts, Mode=OneWay}">
```
is enabled should be bind against both: backup hosts and automatic deletion

davidegiacometti commented:
> I suggest as we have the on/of checkbox for cleanup to set the number boxes to min=1.

The idea is that user can configure retention by:
- Days only: `Days = N - Backups = 0`
- Backups only: `Days = 0 - Backups = N`
- Both: `Days = N - Backups = N`

In case both are set to 0 the warning is displayed and no backups are deleted.

Do you think this makes sense?

htcfreek commented on src/settings-ui/Settings.UI/SettingsXAML/Views/HostsPage.xaml:
```diff
@@ -63,6 +63,83 @@
+                                Background="{ThemeResource SystemFillColorCautionBackgroundBrush}"
+                                ContentAlignment="Vertical"
+                                Visibility="{x:Bind ViewModel.ShowDeleteBackupWarning, Mode=OneWay}">
+                                <InfoBar
```
For a nicer UI: Can we move it as first element in the expander.

htcfreek reviewed (commented):
some ui nits

htcfreek commented:
> > I suggest as we have the on/of checkbox for cleanup to set the number boxes to min=1.
> 
> The idea is that user can configure retention by:
> 
> * Days only: `Days = N - Backups = 0`
> * Backups only: `Days = 0 - Backups = N`
> * Both: `Days = N - Backups = N`
> 
> In case both are set to 0 the warning is displayed and no backups are deleted.
> 
> Do you think this makes sense?

These all feels to complicated. I think keep it simple is the way to go.

And simple for me is:
* Backups are enabled or not.
* Automatic cleanup is "disable" or "based on age" or "based on number". (Can be a drop down.)
* Based on the cleanup type I can configure nothing or the days or the number.

davidegiacometti commented:
> These all feels to complicated. I think keep it simple is the way to go.
> 
> And simple for me is:
> 
> * Backups are enabled or not.
> * Automatic cleanup is "disable" or "based on age" or "based on number". (Can be a drop down.)
> * Based on the cleanup type I can configure nothing or the days or the number.

I started with this idea, but https://github.com/microsoft/PowerToys/issues/37666#issuecomment-2689795329 made me change it.

> Safer would be to keep at least the most recent backup (no matter how old) and then delete others older than 15 days?

But I am happy to have a feedback on this.

davidegiacometti commented:
I need to think about this. The idea around this PR is to remove any hardcoded/hidden logic related to backups.
What do you find confusing of the proposed logic? The warning message?

htcfreek commented:
> I need to think about this: the idea around the PR is to remove any hardcoded/hidden logic related to backups.
> What do you find confusing of the proposed logic? The warning message?

The point that we have multiple ways of doing it which influence each other. And that you can configure it to not clean up even that zhe clean up is enabled.

htcfreek commented:
> ![Animation](https://github.com/user-attachments/assets/24ac15f6-9dc6-4f72-8f1f-d56656f4cb5f)
> 
> @htcfreek any thoughts on this UX?
> - No warning
> - User can set only Days or Copies to 0
> - When Days and Copies are set to 0, "Automatically delete old backups" is disabled and Days and Copies are set to their default

With zero as date or time all backups will be deleted on the `next ????`. We should add description to explain that.

davidegiacometti commented:
The backups cleanup is always executed on editor startup. These options are used to determine for how many days backups are kept or/and how many fixed backups are kept.

htcfreek commented:
> The backups cleanup is always executed on editor startup. These options are used to determine for how many days backups are kept or/and how many fixed backups are kept.

Yes. What I mean is: It makes a difference with 0 days if I start Hosts multiple times per day or not.

htcfreek commented:
> > Yes. What I mean is: It makes a difference with 0 days if I start Hosts multiple times per day or not.
> 
> 0 days means only the number of backups is considered.
> 
> > And for me keeping 0 days means deleting at the day of creating and not keeping indefinitely. That is confusing.
> 
> <img width="593" alt="image" src="https://github.com/user-attachments/assets/2cdf41c2-32de-4792-b91a-6f1f94c3fa79" />
> 
> You mean that this may be confused as `Keep for infinite days AND Keep 10 backups`... I see...

No. As keeping 0 backups for 10 days which means no backups kept.

> TBH I am thinking about reconsider the following solution even if user is forced to choose between number of days or number or backups.
> 
> > * Backups are enabled or not.
> > * Automatic cleanup is "disable" or "based on age" or "based on number". (Can be a drop down.)
> > * Based on the cleanup type I can configure nothing or the days or the number.

Would be see easiest implementation.

davidegiacometti commented:
I think the drop downs are limiting.. 🤔 
I’m going to close this PR for now and try to create a new one with a better approach in the future.
In any case, it is not a priority but rather an enhancement.

davidegiacometti commented:
Hey @htcfreek

Can I have a feedback on these mocks?
Every NumberBox has Minimum set to 1, except for **Based on age | Number of backups to keep** that is optional and has a Minimum of 0.

<img width="770" alt="image" src="https://github.com/user-attachments/assets/9084c944-c17f-47bd-81c3-a8852308607f" />

<img width="776" alt="image" src="https://github.com/user-attachments/assets/349f7a33-355b-4665-b5fe-2abdadb7ccfa" />

<img width="776" alt="image" src="https://github.com/user-attachments/assets/c4571bca-3dd7-40ab-85bd-a9205c3d0948" />

Let me know if you think this solution is clear. If your answer is positive I am going to reopen this PR and tweak the implementation.

htcfreek commented:
@davidegiacometti 

Two suggestions:
- Replace `Number` with `Count` in the drop down.
- For the combined type use `Based on age and count` in the drop down.

> Every NumberBox has Minimum set to 1, except for Based on age | Number of backups to keep that is optional and has a Minimum of 0.

Can't follow you. There are only these two number boxes in the screenshots. Are there any screenshots missing?

And I still see see the problem that 0 means nothing. So we should explain the meaning of 0 in the number box description.

davidegiacometti commented:
Ok for the suggestions.

Please see the update screenshot.
TBH I don't understand what's the problem of 0: see the last NumberBox description.

htcfreek commented:
> TBH I don't understand what's the problem of 0: see the last NumberBox description.
> 

Sory. My mistake. Was confused after all the solutions we discussed.

htcfreek commented:
> Let me know if you think this solution is clear. If your answer is positive I am going to reopen this PR and tweak the implementation.

Yes. It is clear. Let's implement that solution.

davidegiacometti commented:
Thanks for the feedback.. And I like this solution!
Reopening the PR.

htcfreek commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -4895,6 +4895,53 @@ To record a specific window, enter the hotkey with the Alt key in the opposite m
+    <value>Days</value>
+  </data>
+  <data name="Hosts_Backup_CountBased_CountInput.Header" xml:space="preserve">
+    <value>Backups count</value>
```
Isn't the correct grammar "Backup count"?

htcfreek commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -4895,6 +4895,53 @@ To record a specific window, enter the hotkey with the Alt key in the opposite m
+  <data name="Hosts_Backup_AgeCountBased_CountInput.Description" xml:space="preserve">
+    <value>Set an optional number of backups to always retain, despite their age</value>
+  </data>
+  <data name="Hosts_Backup_AgeCountBased_CountInput.Header" xml:space="preserve">
+    <value>Backups count</value>
   </data>
```
Duplicate. Do we really need this twice?

htcfreek commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -4895,6 +4895,53 @@ To record a specific window, enter the hotkey with the Alt key in the opposite m
+  <data name="Hosts_Backup_CountBased_CountInput.Header" xml:space="preserve">
+    <value>Backups count</value>
+  </data>
+  <data name="Hosts_Backup_CountBased_CountInput.Description" xml:space="preserve">
+    <value>Set the number of backups to retain, deleting older ones once the limit is reached</value>
+  </data>
+  <data name="Hosts_Backup_AgeCountBased_DaysInput.Description" xml:space="preserve">
+    <value>Set the number of days to retain backups, deleting older ones once the limit is reached</value>
```
Can we have two independent sentences instead of using a comma? Would be easier to understand.

For me "keep" sounds better than "retain".

davidegiacometti commented:
Addressed the last feedback and updated the screenshot in the first post.
I'll take some time to review and test once again before removing the draft from the PR.

davidegiacometti commented on src/modules/Hosts/Hosts/Settings/UserSettings.cs:
```diff
@@ -54,6 +65,11 @@ public UserSettings()
             AdditionalLinesPosition = HostsAdditionalLinesPosition.Top;
             Encoding = HostsEncoding.Utf8;
+            BackupHosts = true;
+            BackupPath = @"C:\Windows\system32\drivers\etc";
```
Nice catch!
I used the same logic we are using for the hosts file path: no issues so far, so I think a fallback isn't needed.
https://github.com/microsoft/PowerToys/blob/c6776d0d45009a7c1fdbb1360ef712fa789312e9/src/modules/Hosts/HostsUILib/Helpers/HostsService.cs#L53

davidegiacometti commented:
Don't worry and thanks for the review.
I have pushed the changes.

davidegiacometti commented:
/azp run

yeelam-gordon commented on src/modules/Hosts/HostsUILib/Helpers/BackupManager.cs:
```diff
@@ -0,0 +1,107 @@
+            DeleteAll(backups);
+        }
+
+        private IEnumerable<IFileInfo> GetAll()
```
`GetAll` assumes the backup folder still exists. Once a user points the setting at a removable drive or deletes the directory, `Directory.GetFiles` throws, and the cleanup thread now logs "Failed to delete backup" every launch without actually removing anything. Please guard for a missing folder (e.g. return an empty sequence when `_fileSystem.Directory.Exists` is false) so cleanup becomes a no-op instead of an error flood.

vanzue reviewed (approved):
See pipeline passes, approve for unblocking
